### PR TITLE
Revert frozen training arguments

### LINKED ISSUE #25897: TrainingArguments frozen: leads to issues in Trainer
### System Info

- `transformers` version: 4.32.1
- Platform: Linux-5.14.0-284.25.1.el9_2.x86_64-x86_64-with-glibc2.34
- Python version: 3.11.4
- Huggingface_hub version: 0.16.4
- Safetensors version: 0.3.1
- Accelerate version: 0.22.0
- Accelerate config:    not found
- PyTorch version (GPU?): 2.0.1+cu117 (True)

### Who can help?

@muellerz and @pacman100

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

I recently encountered an annoying issue when trying to use the wandb integration for hyperparameter search. When doing so, the trainer will try to reset some of its `args` (TrainingArguments). But it seems that since [recently](https://github.com/huggingface/transformers/commit/ca51499248b986ebf3991848234ef2d8bc81a36a), the choice was made to freeze training arguments.

So this won't work.

```python
from transformers import TrainingArguments
args = TrainingArguments(output_dir="test")
args.output_dir = "test_dir"
args.run_name = "hello"
```

### Expected behavior

As a whole I am not a fan of this commit to freezing a dataclass like that - which is not done in the canonical Python way of using `@dataclass(frozen=True)` so it is hard to spot if you're reading to code. 

If it _must_ stay, then some more changes need to be made throughout the lib. Such as this line

https://github.com/huggingface/transformers/blob/0f08cd205a440d23e6bf924cddd73ff48e09fe35/src/transformers/integrations/integration_utils.py#L758

Should probably be

```python
args = dataclasses.replace(args, run_name=None)
```

(untested)

Currently I do not have the time for a PR but this could be a good first issue for someone who would like to contribute.

### PR DESCRIPTION
**Pull request by muellerzr**
# What does this PR do?

This PR reverts a recent change that made the `TrainingArguments` immutable objects. While this may have been our *intended* design decision, as users were already integrated with modifying the training_args directly, we've decided to revert this change.

Users code (if already modified to deal with this issue) does not actually need changing, since it uses dataclass utils under the hood. However they can change any cases of:

```python
training_args = dataclasses.replace(training_args, item_a=value_a)
```

With:
```python
training_args.item_a = value_a
```

Fixes # (issue)

Solves https://github.com/huggingface/transformers/issues/25897

Linking a few known repos/issues/PR's that reference this:

- https://github.com/huggingface/trl/pull/676
- https://github.com/huggingface/optimum-intel/pull/412
- https://github.com/artidoro/qlora/issues/253
- https://github.com/tomaarsen/SpanMarkerNER/pull/27

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@amyeroberts @LysandreJik

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
Looks good to me, cc @ArthurZucker who's watching today

File: src/transformers/trainer.py
Comment (ArthurZucker) [inline src/transformers/trainer.py:1]: Guessing this is from #25540

**Review by ArthurZucker — APPROVED** [review]
Looks good. Changes from the PR were all reverted. I can't know better than you if there was another PR related (like #25540) so make sure to link them all! 
Would be worth adding a small doc mentioning maybe the expected *bugs*  you could have because they are not frozen?

File: src/transformers/trainer.py
Comment (muellerzr) [reply src/transformers/trainer.py:1]: Yep!

**Review by ArthurZucker — APPROVED** [review]
Thanks! 😉

**Pull request merged**